src/instrument.cpp

```diff
@@ -282,15 +286,82 @@ void BytecodeRewriter::rewriteCode() {
     put16(max_locals);
 
     u32 code_length = get32();
-    put32(code_length + EXTRA_BYTECODES);
+    const u8* code = get(code_length);
+    u32 code_length_idx = _dst_len;
+    put32(code_length); // to be fixed later
+
+    int code_segment_begin = _dst_len;
+
+    u32 max_relocation = 0;
+    for (u32 i = 0; i < code_length;) {
+        u8 opcode = code[i];
+
+        if (opcode == 0xb0 || opcode == 0xaf || opcode == 0xae || opcode == 0xac ||
+            opcode == 0xad || opcode == 0xb1 || opcode == 0xbf) {
+            max_relocation += EXTRA_BYTECODES;
+        } else if (opcode == 0xa8 || opcode == 0xa7) {
+            max_relocation += 2;
+        }
+        i += OPCODE_LENGTH[opcode];
+    }
+
+    // For each index in the (original) bytecode, this holds the rightwards offset
+    // in the modified bytecode.
+    u32 relocation_table[code_length];
+    u32 current_relocation = 0;
+    std::vector<u32> jumps;
+    for (u32 i = 0; i < code_length;) {
+        u8 opcode = code[i];
+
+        relocation_table[i] = 0;
+        if (opcode == 0xa7 || opcode == 0xa8) {
+            jumps.push_back(i);
+
+            int16_t offset = (int16_t) ntohs(*(u16*)(code + i + 1));
+            if (offset > std::numeric_limits<int16_t>::max() - max_relocation) {
+                current_relocation += 2;
+
+                put8(opcode == 0xa8 ? 0xc9 : 0xc8);
+                put16(0); // to be fixed later
+                put16(offset);
+            }
+            relocation_table[i+1] = current_relocation;
+            relocation_table[i+2] = current_relocation;
 
-    // invokestatic "one/profiler/Instrument.recordSample()V"
-    // nop ensures that tableswitch/lookupswitch needs no realignment
-    put8(0xb8);
-    put16(_cpool_len);
-    put8(0);
-    // The rest of the code is unchanged
-    put(get(code_length), code_length);
+            i += 3;
+            continue;
+        }
+
+        if (opcode == 0xb0 || opcode == 0xaf || opcode == 0xae || opcode == 0xac ||
+            opcode == 0xad || opcode == 0xb1 || opcode == 0xbf) {
```

Comment:
`if ((opcode >= JVM_OPC_ireturn && opcode <= JVM_OPC_return) || opcode == JVM_OPC_athrow)`

in all similar places